### PR TITLE
Refine rename functions in datalake

### PR DESCRIPTION
**Pull request by katmsft**
# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/master/CONTRIBUTING.md#pull-requests).

- [x] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [x] Doxygen docs
- [x] Unit tests
- [x] No unwanted commits/changes
- [x] Descriptive title/description
  - [x] PR is single purpose
  - [x] Related issue listed
- [x] Comments in source
- [x] No typos
- [x] Update changelog
- [x] Not work-in-progress
- [x] External references or docs updated
- [x] Self review of PR done
- [x] Any breaking changes?

### CODE REVIEW

**Comment by katmsft** [comment]
@tg-msft Please take a look at the rename API.
Several things:
1. The returned type is still a `result` because rename is a pageable API, customer may not be able to finish the rename operation in a single call when renaming a directory that contains many subdirectories and files. However, the client is also returned in the result type. It should not be a problem for renaming a file, but I adapted the same pattern to align with renaming a directory for consistency.
2. The rename API for the directory is having a `SinglePage` suffix, to mark that it is a pageable API and might need further enumerations.
3. As agreed, the rename API only exists in the parent resource, as in the `DataLakeFileSystemClient` can rename file/directory, and `DataLakeDirectoryClient` can rename file/subdirectory. File cannot be renamed with its own client.
4. `Models::PathRenameMode Mode` is supported by REST, I don't know if it is a missing feature for .NET but C++ left it as is. Please share the reasoning of removing it if preferred.

**Comment by tg-msft** [comment]
I think your choices make sense and all of that sounds good to me.  @seanmcc-msft for the question about PathRenameMode in .NET.

**Comment by seanmcc-msft** [comment]
> I think your choices make sense and all of that sounds good to me. @seanmcc-msft for the question about PathRenameMode in .NET.

We only support the `Legancy` Rename Mode in .NET, Java, and Python.  This is because in `POSIX` mode, we don't know what the destination will be without making a second rest call.  In `POSIX` mode, if the destination exists, the path you are trying to rename will be placed under the existing destination.  If not, it is placed at the destination where you expect it.

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_directory_client.hpp [hunk 1]
```diff
@@ -115,21 +115,39 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
...
+     * @param subdirectoryName The subdirectory that gets renamed.
+     * @param destinationDirectoryPath The destinationPath the source subdirectory is renaming to.
+     * @param options Optional parameters to rename a directory.
+     * @return Azure::Core::Response<Models::RenameDataLakeSubdirectorySinglePageResult> containing
+     * the information returned when renaming the directory, including the DataLakeDirectoryClient
+     * that's targeting the renamed directory.
      * @remark This request is sent to dfs endpoint.
      */
-    Azure::Core::Response<Models::RenameDataLakeDirectoryResult> Rename(
+    Azure::Core::Response<Models::RenameDataLakeSubdirectorySinglePageResult>
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_directory_client.hpp:145]: I prefer this returns a `DataLakeDirectoryClient`

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp [hunk 2]
```diff
@@ -258,6 +258,52 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
...
+     *        rename operation to continue renaming the directory.
+     */
+    Azure::Core::Nullable<std::string> ContinuationToken;
+
+    /**
+     * @brief This parameter determines the behavior of the rename operation. The value must be
+     *        PathRenameMode::Legacy or PathRenameMode::Posix, and the default value will be
+     *        PathRenameMode::Posix.
+     */
+    Models::PathRenameMode Mode = Models::PathRenameMode::Posix;
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp:288]: you may want to reconsider the default value given what Sean said https://github.com/Azure/azure-sdk-for-cpp/pull/1493#issuecomment-768761597

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_file_system_client.hpp [hunk 3]
```diff
@@ -201,6 +201,40 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
...
+     * @param destinationDirectoryPath The destinationPath the source directory is renaming to.
+     * @param options Optional parameters to rename a directory.
+     * @return Azure::Core::Response<Models::RenameDataLakeDirectorySinglePageResult> containing the
+     * information returned when renaming the directory, including the DataLakeDirectoryClient
+     * targeting the renamed directory.
+     * @remark This request is sent to dfs endpoint.
+     */
+    Azure::Core::Response<Models::RenameDataLakeDirectorySinglePageResult>
+    RenameDirectorySinglePage(
+        const std::string& directoryName,
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_file_system_client.hpp:233]: Where is that `recursive` option?

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_file_system_client.hpp [hunk 4]
```diff
@@ -201,6 +201,40 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
...
+     *        if the destination already exists and has a lease the lease is broken.
+     * @param directoryName The directory that gets renamed.
+     * @param destinationDirectoryPath The destinationPath the source directory is renaming to.
+     * @param options Optional parameters to rename a directory.
+     * @return Azure::Core::Response<Models::RenameDataLakeDirectorySinglePageResult> containing the
+     * information returned when renaming the directory, including the DataLakeDirectoryClient
+     * targeting the renamed directory.
+     * @remark This request is sent to dfs endpoint.
+     */
+    Azure::Core::Response<Models::RenameDataLakeDirectorySinglePageResult>
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_file_system_client.hpp:231]: I prefer it returns a `DataLakeClient`

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp (hunk 2 above)
Comment (seanmcc-msft) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp:288]: I would recommend removing this parameter entirely.  We can add it back in the future if customers ask for it, but that has not been the case on the other platforms.

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp:288]: > I would recommend removing this parameter entirely. We can add it back in the future if customers ask for it, but that has not been the case on the other platforms.


I agree

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_file_system_client.hpp (hunk 3 above)
Comment (seanmcc-msft) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_file_system_client.hpp:233]: Recursive should be set to `true` for directories by the SDK, the customer shouldn't have to worry about it.

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_file_system_client.hpp:233]: > Recursive should be set to true for directories by the SDK, the customer shouldn't have to worry about it.

Yeah, that's what I'm thinking. @katmsft please confirm `recursive` is set to `true`.

Comment (katmsft) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_file_system_client.hpp:233]: There is no such option, in both .NET and in REST: https://github.com/Azure/azure-sdk-for-net/blob/44e9c2b2190c03ec15862eeb101eb37fde573e08/sdk/storage/Azure.Storage.Files.DataLake/src/DataLakePathClient.cs#L1832
https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/path/create

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp (hunk 2 above)
Comment (katmsft) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp]: Removed.

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_directory_client.hpp (hunk 1 above)
Comment (katmsft) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_directory_client.hpp:145]: How would you also return the `ContinuationToken` then?

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_file_system_client.hpp (hunk 4 above)
Comment (katmsft) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_file_system_client.hpp:231]: How would you also return the `ContinuationToken` then?

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp [hunk 5]
```diff
@@ -258,6 +258,45 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
...
+    Azure::Core::Context Context;
+
+    /**
+     * @brief When renaming a directory, the number of paths that are renamed with each
+     *        invocation is limited. If the number of paths to be renamed exceeds this limit,
+     *        a continuation token is returned in this response header. When a continuation token
+     *        is returned in the response, it must be specified in a subsequent invocation of the
+     *        rename operation to continue renaming the directory.
+     */
+    Azure::Core::Nullable<std::string> ContinuationToken;
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp:281]: `ContinuationToken` should be removed?

File: sdk/storage/azure-storage-files-datalake/src/datalake_file_system_client.cpp [hunk 6]
```diff
@@ -395,4 +395,21 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
         std::move(ret), result.ExtractRawResponse());
   }
 
+  Azure::Core::Response<DataLakeFileClient> DataLakeFileSystemClient::RenameFile(
+      const std::string& fileName,
+      const std::string& destinationFilePath,
+      const RenameDataLakeFileOptions& options) const
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/src/datalake_file_system_client.cpp:401]:
```suggestion
      const RenameDataLakeFileOptions& options = RenameDataLakeFileOptions()) const
```

File: sdk/storage/azure-storage-files-datalake/src/datalake_file_system_client.cpp [hunk 7]
```diff
@@ -395,4 +395,21 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
...
+  {
+    return this->GetDirectoryClient("").RenameFile(fileName, destinationFilePath, options);
+  }
+
+  Azure::Core::Response<DataLakeDirectoryClient> DataLakeFileSystemClient::RenameDirectory(
+      const std::string& directoryName,
+      const std::string& destinationDirectoryPath,
+      const RenameDataLakeDirectoryOptions& options) const
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/src/datalake_file_system_client.cpp:409]:
```suggestion
      const RenameDataLakeDirectoryOptions& options = RenameDataLakeDirectoryOptions()) const
```

**Review by Jinming-Hu — APPROVED** [review]

**Pull request merged**